cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/queue/QueueReaderFactory.java

```diff
@@ -32,15 +35,24 @@
  */
 public final class QueueReaderFactory {
 
+  private final AuthorizationEnforcer authorizationEnforcer;
+  private final AuthenticationContext authenticationContext;
+
+  @Inject
+  public QueueReaderFactory(AuthorizationEnforcer authorizationEnforcer, AuthenticationContext authenticationContext) {
+    this.authorizationEnforcer = authorizationEnforcer;
+    this.authenticationContext = authenticationContext;
+  }
+
   public <T> QueueReader<T> createQueueReader(Supplier<QueueConsumer> consumerSupplier,
                                               int batchSize, Function<ByteBuffer, T> decoder) {
     return new SingleQueue2Reader<>(consumerSupplier, batchSize, decoder);
   }
 
   public <T> QueueReader<T> createStreamReader(Supplier<StreamConsumer> consumerSupplier,
-                                               int batchSize, Function<StreamEvent, T> transformer,
-                                               AuthenticationContext authenticationContext,
-                                               AuthorizationEnforcer authorizationEnforcer) {
+                                               int batchSize, Function<StreamEvent, T> transformer) throws Exception {
+    StreamId streamId = consumerSupplier.get().getStreamId().toEntityId();
```

Comment:
calling `consumerSupplier.get()` will create a new instance of `StreamConsumer`, which requires explicitly closing. Can we modify this method so that the caller provides the `StreamId` as well?